File: frontend/src/services/bookingService.js
// Booking API calls.
import api from './api.js';

export const createBooking = ({ slot_id, start_time, end_time }) =>
  api.post('/bookings/create', {
    slot_id,
    ...(start_time ? { start_time: new Date(start_time).toISOString() } : {}),
    ...(end_time ? { end_time: new Date(end_time).toISOString() } : {}),
  });

export const myBookings = () => api.get('/bookings/my-bookings');
export const getBooking = (id) => api.get(`/bookings/${id}`);
export const cancelBooking = (id) => api.post(`/bookings/${id}/cancel`);

// Gate staff (admin/operator): scan result → check-in / check-out.
export const verifyQr = (qrData) => api.post('/bookings/verify-qr', { qr_data: qrData });

File: mobile/lib/services/booking_service.dart
import '../models/booking.dart';
import 'api_service.dart';

class QrVerifyResult {
  final String direction; // 'check-in' | 'check-out'
  final Booking booking;
  const QrVerifyResult(this.direction, this.booking);
}

/// Booking API calls (incl. the operator gate-scan).
class BookingService {
  final _api = ApiService.instance;

  Future<Booking> createBooking({
    required String slotId,
    DateTime? startTime,
    DateTime? endTime,
  }) async {
    final body = <String, dynamic>{'slot_id': slotId};
    if (startTime != null) body['start_time'] = startTime.toUtc().toIso8601String();
    if (endTime != null) body['end_time'] = endTime.toUtc().toIso8601String();
    final data = await _api.post('/bookings/create', body: body);
    return Booking.fromJson(data['booking'] as Map<String, dynamic>);
  }

  Future<List<Booking>> myBookings() async {
    final data = await _api.get('/bookings/my-bookings');
    return (data['bookings'] as List<dynamic>)
        .map((b) => Booking.fromJson(b as Map<String, dynamic>))
        .toList();
  }

  Future<Booking> cancelBooking(String id) async {
    final data = await _api.post('/bookings/$id/cancel');
    return Booking.fromJson(data['booking'] as Map<String, dynamic>);
  }

  /// Gate scan (admin/operator). Backend decides check-in vs check-out.
  Future<QrVerifyResult> verifyQr(String qrData) async {
    final data = await _api.post('/bookings/verify-qr', body: {'qr_data': qrData});
    return QrVerifyResult(
      data['direction'] as String,
      Booking.fromJson(data['booking'] as Map<String, dynamic>),
    );
  }
}
